examples/with-apollo/lib/apollo.js

```diff
@@ -0,0 +1,130 @@
+import React, { useMemo } from 'react'
+import Head from 'next/head'
+import { ApolloProvider } from '@apollo/react-hooks'
+import { ApolloClient, InMemoryCache, HttpLink } from 'apollo-boost'
+import fetch from 'isomorphic-unfetch'
+
+let apolloClient = null
+
+/**
+ * Creates and provides the apolloContext
+ * to a next.js PageTree. Use it by wrapping
+ * your PageComponent via HOC pattern.
+ * @param {Function|Class} PageComponent
+ * @param {Object} [config]
+ * @param {Boolean} [config.ssr=true]
+ */
+export function withApollo (PageComponent, { ssr = true } = {}) {
+  const WithApollo = ({ apolloClient, apolloState, ...pageProps }) => {
+    const client = useMemo(
+      () => apolloClient || initApolloClient(apolloState),
+      []
+    )
+    return (
+      <ApolloProvider client={client}>
+        <PageComponent {...pageProps} />
+      </ApolloProvider>
+    )
+  }
+
+  // Set the correct displayName in development
+  if (process.env.NODE_ENV !== 'production') {
+    const displayName =
+      PageComponent.displayName || PageComponent.name || 'Component'
+
+    if (displayName === 'App') {
+      console.warn('This withApollo HOC only works with PageComponents.')
+    }
+
+    WithApollo.displayName = `withApollo(${displayName})`
+  }
+
+  // Allow Next.js to remove getInitialProps from the browser build
+  if (typeof window === 'undefined') {
+    if (ssr) {
+      WithApollo.getInitialProps = async ctx => {
+        const { AppTree } = ctx
+
+        let pageProps = {}
+        if (PageComponent.getInitialProps) {
+          pageProps = await PageComponent.getInitialProps(ctx)
```

Comment:
Should this be ctx.ctx, because ctx is AppContext here, and the page expects a PageContext?

Reply:
nope. This PR uses an PageHoc instead of an AppHoc. So we have PageContext here now. Nevertheless, thanks a lot for having a look!

Reply:
Ah that's right, thanks!